File: rust/src/bin/structual-patterns/facade.rs
use std::ops::{SubAssign};

struct Account {
    name: String
}

impl Account {
    fn new(name: &str) -> Self {
        Self {
            name: name.to_string()
        }
    }

    fn check_name(&self, name: &str) -> bool {
        self.name == name
    }
}

struct Wallet {
    balance: f32
}

impl Wallet {
    fn new() -> Self {
        Self {
            balance: 0.0
        }
    }

    fn deposit(&mut self, amount: f32) {
        self.balance += amount;
    }

    fn withdraw(&mut self, amount: f32) {
        self.balance.sub_assign(amount);
    }

    fn get_balance(&self) -> f32 {
        self.balance
    }
}

struct SecurityCode {
    code: String
}

impl SecurityCode {
    fn new(code: &str) -> Self {
        Self {
            code: code.to_string()
        }
    }

    fn check_code(&self, code: &str) -> bool {
        self.code == code
    }
}

struct Notification;

impl Notification {
    fn notify(&self, message: &str) {
        println!("Notification: {}", message);
    }
}

struct WalletFacade {
    account: Account,
    wallet: Wallet,
    security_code: SecurityCode,
    notification: Notification
}

impl WalletFacade {
    fn new(name: &str, code: &str) -> Self {
        Self {
            account: Account::new(name),
            wallet: Wallet::new(),
            security_code: SecurityCode::new(code),
            notification: Notification
        }
    }

    fn deposit(&mut self, name: &str, code: &str, amount: f32) {
        if self.account.check_name(name) && self.security_code.check_code(code) {
            self.wallet.deposit(amount);
            self.notification.notify(&format!("Deposited {} to {}'s wallet.", amount, name));
        } else {
            println!("Invalid account name or security code.");
        }
    }

     fn withdraw(&mut self, name: &str, code: &str, amount: f32) {
        if self.account.check_name(name) && self.security_code.check_code(code) {
            if self.wallet.get_balance() >= amount {
                self.wallet.withdraw(amount);
                self.notification.notify(&format!("Withdrew {} from {}'s wallet.", amount, name));
            } else {
                println!("Insufficient balance.");
            }
        } else {
            println!("Invalid account name or security code.");
        }
     }

     fn get_balance(&self, name: &str, code: &str) -> f32 {
        if self.account.check_name(name) && self.security_code.check_code(code) {
            self.wallet.get_balance()
        } else {
            println!("Invalid account name or security code.");
            0.0
        }
    }
}

fn main() {
   let mut wallet = WalletFacade::new("John", "1234");

    wallet.deposit("John", "1234", 100.0);
    wallet.deposit("John", "2222", 20.0);
    wallet.withdraw("John", "1234", 10.0);
    println!("Current balance: {}", wallet.get_balance("John", "1234"));
}

File: typescript/structual-patterns/facade.ts
class Account {
  private name: string;

  constructor(name: string) {
    this.name = name;
  }

  checkName(name: string): boolean {
    return this.name === name;
  }
}

class Wallet {
  private balance: number;

  constructor() {
    this.balance = 0;
  }

  deposit(amount: number): void {
    this.balance += amount;
  }

  withdraw(amount: number): void {
    if (this.balance >= amount) {
      this.balance -= amount;
    } else {
      console.log("Insufficient balance");
    }
  }

  getBalance(): number {
    return this.balance;
  }
}

class SecurityCode {
  private code: string;

  constructor(code: string) {
    this.code = code;
  }

  checkCode(code: string): boolean {
    return this.code === code;
  }
}

class MyNotification {
  notify(message: string): void {
    console.log(`Notification: ${message}`);
  }
}

class WalletFacade {
  private account: Account;
  private wallet: Wallet;
  private securityCode: SecurityCode;
  private notification: MyNotification;

  constructor(name: string, code: string) {
    this.account = new Account(name);
    this.wallet = new Wallet();
    this.securityCode = new SecurityCode(code);
    this.notification = new MyNotification();
  }

  deposit(name: string, code: string, amount: number): void {
    if (this.account.checkName(name) && this.securityCode.checkCode(code)) {
      this.wallet.deposit(amount);
      this.notification.notify(`Deposited ${amount} to ${name}'s wallet.`);
    } else {
      console.log("Invalid account name or security code.");
    }
  }

  withdraw(name: string, code: string, amount: number): void {
    if (this.account.checkName(name) && this.securityCode.checkCode(code)) {
      this.wallet.withdraw(amount);
      this.notification.notify(`Withdrew ${amount} from ${name}'s wallet.`);
    } else {
      console.log("Invalid account name or security code.");
    }
  }

  getBalance(name: string, code: string): number {
    if (this.account.checkName(name) && this.securityCode.checkCode(code)) {
      return this.wallet.getBalance();
    } else {
      console.log("Invalid account name or security code.");
      return 0;
    }
  }
}

function clientCode(wallet: WalletFacade) {
  wallet.deposit("John", "1234", 100);
  wallet.deposit("John", "2222", 20);

  wallet.withdraw("John", "1234", 50);

  console.log(`Current balance: ${wallet.getBalance("John", "1234")}`);
}

const wallet = new WalletFacade("John", "1234");
clientCode(wallet);
